en: add backup&restore to gcs using br (#582)
pingcap/docs-tidb-operator: en/backup-to-gcs-using-br.md

---
title: Back up Data to GCS Using BR
summary: Learn how to back up data to Google Cloud Storage (GCS) using BR.
---

# Back up Data to GCS Using BR

> This document describes how to back up the data of a TiDB cluster in Kubernetes to [Google Cloud Storage](https://cloud.google.com/storage/docs/) (GCS). "Backup" in this document refers to full backup (ad-hoc full backup and scheduled full backup). [BR](https://pingcap.com/docs/stable/br/backup-and-restore-tool/) is used to get the logic backup of the TiDB cluster, and then this backup data is sent to GCS.

Reviewer: ```suggestion
This document describes how to back up the data of a TiDB cluster in Kubernetes to [Google Cloud Storage](https://cloud.google.com/storage/docs/) (GCS). "Backup" in this document refers to full backup (ad-hoc full backup and scheduled full backup). [BR](https://pingcap.com/docs/stable/br/backup-and-restore-tool/) is used to get the backup of the TiDB cluster, and then the backup data is sent to GCS.
```

The backup method described in this document is implemented using Custom Resource Definition (CRD) in TiDB Operator v1.1 or later versions.

## Ad-hoc full backup

> Ad-hoc full backup describes the backup by creating a `Backup` Custom Resource (CR) object. TiDB Operator performs the specific backup operation based on this `Backup` object. If an error occurs during the backup process, TiDB Operator does not retry, and you need to handle this error manually.

Reviewer 2: `you need` may be change to `customer need` ?

Reviewer: I think the current expression is OK, the users are reading this doc.

> Currently, the above three authorization methods are supported for the ad-hoc full backup. This document provides examples in which the data of the `demo1` TiDB cluster in the `test1` Kubernetes namespace is backed up to AWS storage and all the above methods are used in the examples.

Reviewer: This sentence does not match with the Chinese version.

    apiVersion: pingcap.com/v1alpha1
    kind: Backup
    metadata:
    name: demo1-backup-gcs
    namespace: test1
    spec:
      # backupType: full
      from:
        host: ${tidb-host}
        port: ${tidb-port}
        user: ${tidb-user}
        secretName: backup-demo1-tidb-secret
      br:
        cluster: demo1
        clusterNamespace: test1
        # logLevel: info
        # statusAddr: ${status-addr}
        # concurrency: 4
        # rateLimit: 0
        # checksum: true
        # sendCredToTikv: true
      gcs:
        projectId: ${project-id}
        secretName: gcs-secret
        bucket: ${bucket}
        prefix: ${prefix}
        # location: us-east1
        # storageClass: STANDARD_IA
        # objectAcl: private
    ```

    In the example above, some parameters in `spec.br` can be ignored, such as `logLevel`, `statusAddr`, `concurrency`, `rateLimit`, `checksum`, and `sendCredToTikv`.

    <details>
    <summary>More description of parameters</summary>

    - `spec.br.cluster`: The name of the cluster to be backed up.
    - `spec.br.clusterNamespace`: The `namespace` of the cluster to be backed up.
    - `spec.br.logLevel`: The log level (`info` by default).

>     - `spec.br.statusAddr`: The HTTP port used for listening a process status in BR to facilitate debugging. If not specified, BR does not listen the process status by default.

Reviewer: ```suggestion
    - `spec.br.statusAddr`: The listening address through which BR provides statistics. If not specified, BR does not listen on any status address by default.
```

Reviewer: From https://docs.pingcap.com/tidb/stable/backup-and-restore-tool#common-options

    </details>

    This example backs up all data in the TiDB cluster to GCS. Some parameters in `spec.gcs` can be ignored, such as `location`, `objectAcl`, and `storageClass`.

    The `projectId` in the configuration is the unique identifier for the user project on GCP. For how to obtain the identifier, see [GCP Documentation](https://cloud.google.com/resource-manager/docs/creating-managing-projects).

    <details>
    <summary>Configure <code>storageClass</code>.</summary>

    GCS supports the following `storageClass` types:

    * `MULTI_REGIONAL`
    * `REGIONAL`
    * `NEARLINE`
    * `COLDLINE`
    * `DURABLE_REDUCED_AVAILABILITY`

    If you do not configure `storageClass`, the default is `COLDLINE`. See [GCS Documentation](https://cloud.google.com/storage/docs/storage-classes) for details.
    
    </details>

    <details>
    <summary>Configure the object access-control list (ACL) strategy</summary>

    CS supports the following ACL strategies:

    * `authenticatedRead`
    * `bucketOwnerFullControl`
    * `bucketOwnerRead`
    * `private`
    * `projectPrivate`
    * `publicRead`

>     If you do not configure `storageClass`, the default is `COLDLINE`. See [GCS Documentation](https://cloud.google.com/storage/docs/storage-classes) for details.

Reviewer: This needs to update.

    </details>

2. After creating the `Backup` CR, use the following command to check the backup status:

    {{< copyable "shell-regular" >}}

    ```shell
    kubectl get bk -n test1 -owide
    ```

<details>

<summary>More descriptions of fields in the <code>Backup</code> CR</summary>

> * `.spec.metadata.namespace`: The namespace where the `Backup` CR is located.

Reviewer 3: Some content is missing in your translation:

* `.spec.tikvGCLifeTime`：备份中的临时 `tikv_gc_lifetime` 时间设置，默认为 72h。

    在备份开始之前，若 TiDB 集群的 `tikv_gc_lifetime` 小于用户设置的 `spec.tikvGCLifeTime`，为了保证备份的数据不被 TiKV GC 掉，TiDB Operator 会在备份前[调节 `tikv_gc_lifetime`](https://docs.pingcap.com/zh/tidb/stable/dumpling-overview#导出大规模数据时的-tidb-gc-设置) 为 `spec.tikvGCLifeTime`。

    备份结束后不论成功或者失败，只要老的 `tikv_gc_lifetime` 比设置的 `.spec.tikvGCLifeTime` 小，TiDB Operator 都会尝试恢复 `tikv_gc_lifetime` 为备份前的值。在极端情况下，TiDB Operator 访问数据库失败会导致 TiDB Operator 无法自动恢复 `tikv_gc_lifetime` 并认为备份失败。

    此时，可以通过下述语句查看当前 TiDB 集群的 `tikv_gc_lifetime`：

    ```sql
    select VARIABLE_NAME, VARIABLE_VALUE from mysql.tidb where VARIABLE_NAME like "tikv_gc_life_time";
    ```

    如果发现 `tikv_gc_lifetime` 值过大（通常为 10m），则需要按照[调节 `tikv_gc_lifetime`](https://docs.pingcap.com/zh/tidb/stable/dumpling-overview#导出大规模数据时的-tidb-gc-设置) 将 `tikv_gc_lifetime` 调回原样。

* `.spec.cleanData`：设置为 true 时删除该 Backup CR 时会同时清除该 CR 备份出的数据，默认为 false。值得注意的是，在 v1.1.2 以及之前版本不存在该字段，且默认在删除 CR 的同时删除备份的文件。若 v1.1.3 及之后版本的用户希望保持该行为，需要设置该字段为 true。

> * `.spec.from.host`: The address of the TiDB cluster to be backed up.

Reviewer 3: ```为需要导出的 TiDB 的 service name，例如 `basic-tidb`。``` is missing.

* `.spec.from.port`: The port of the TiDB cluster to be backed up.
* `.spec.from.user`: The accessing user of the TiDB cluster to be backed up.
* `.spec.gcs.bucket`: The name of the bucket which stores data.

> * `.spec.gcs.prefix`: This field is used to make up the path of the remote storage: `s3://${.spec.gcs.bucket}/${.spec.gcs.prefix}/backupName`. This field can be ignored.

Reviewer: Should be `gcs://`, @shuijing198799 please confirm the format.

Reviewer 2: yes here should be `gcs://`

> * `.spec.from.tidbSecretName`: The secret of the user password of the `.spec.from.user` TiDB cluster.

Reviewer: ```suggestion
* `.spec.from.tidbSecretName`: The secret containg the password of the `.spec.from.user` in the TiDB cluster.
```

    * `COLDLINE`
    * `DURABLE_REDUCED_AVAILABILITY`

    If you do not configure `storageClass`, the default type is `COLDLINE`. See [GCS Documentation](https://cloud.google.com/storage/docs/storage-classes) for details.

    </details>

    <details>
    <summary>Configure the object access-control list (ACL) strategy</summary>

    GCS supports the following ACL strategies:

    * `authenticatedRead`
    * `bucketOwnerFullControl`
    * `bucketOwnerRead`
    * `private`
    * `projectPrivate`
    * `publicRead`

    If you do not configure the object ACL strategy, the default strategy is `private`. See [GCS Documentation](https://cloud.google.com/storage/docs/access-control/lists) for details.
    </details>

2. After creating the `Backup` CR, use the following command to check the backup status:

    {{< copyable "shell-regular" >}}

    ```shell
    kubectl get bk -n test1 -owide
    ```

<details>

<summary>More descriptions of fields in the <code>Backup</code> CR</summary>

* `.spec.metadata.namespace`: The namespace where the `Backup` CR is located.
* `.spec.tikvGCLifeTime`: The temporary `tikv_gc_lifetime` time setting during the backup. Defaults to 72h.

>     Before the backup begins, if the `tikv_gc_lifetime` in the TiDB cluster is smaller than the `spec.tikvGCLifeTime` set by the user, TiDB Operator adjusts `tikv_gc_lifetime` to `spec.tikvGCLifeTime`. This operation makes sure that the backup data is not garbage collected by TiKV.

Author: The related link (dumpling-overview.md) is not available in English yet.
I'll add those link when this PR (https://github.com/pingcap/docs-cn/pull/3800) is translated.

Reviewer: I think we can copy the command here, and no need to link to that doc.

>     Before the backup begins, if the `tikv_gc_lifetime` in the TiDB cluster is smaller than the `spec.tikvGCLifeTime` set by the user, TiDB Operator adjusts `tikv_gc_lifetime` to `spec.tikvGCLifeTime`. This operation makes sure that the backup data is not garbage collected by TiKV.

Reviewer 3: ```suggestion
    Before the backup begins, if the `tikv_gc_lifetime` setting in the TiDB cluster is smaller than `spec.tikvGCLifeTime` set by the user, TiDB Operator adjusts the value of `tikv_gc_lifetime` to the value of `spec.tikvGCLifeTime`. This operation makes sure that the backup data is not garbage-collected by TiKV.
```

>     After the backup, whether the backup is successful or not, if the previous `tikv_gc_lifetime` is smaller than `.spec.tikvGCLifeTime`, TiDB Operator will try to set `tikv_gc_lifetime` to the previous value.

Reviewer 3: ```suggestion
    After the backup, no matter whether the backup is successful or not, as long as the previous `tikv_gc_lifetime` is smaller than `.spec.tikvGCLifeTime`, TiDB Operator will try to set `tikv_gc_lifetime` to the previous value.
```

>     In extreme cases, if TiDB Operator fails to access the database, TiDB Operator cannot automatically recover the value of `tikv_gc_lifetime` and thinks the backup fails. At this time, you can view `tikv_gc_lifetime` of the current TiDB cluster using the following statement:

Reviewer 3: ```suggestion
    In extreme cases, if TiDB Operator fails to access the database, TiDB Operator cannot automatically recover the value of `tikv_gc_lifetime` and treats the backup as failed. At this time, you can view `tikv_gc_lifetime` of the current TiDB cluster using the following statement:
```

    {{< copyable "sql" >}}

    ```sql
    select VARIABLE_NAME, VARIABLE_VALUE from mysql.tidb where VARIABLE_NAME like "tikv_gc_life_time";
    ```

    If the value of `tikv_gc_lifetime` is set too large (usually 10m), you need to set it back to the previous value.

> * `.spec.cleanPolicy`: The backup file clean policy when the backup CR is deleted after the backup is completed.

Reviewer 3: ```suggestion
* `.spec.cleanPolicy`: The clean policy of the backup file when the backup CR is deleted after the backup is completed.
```

    Three clean policies are supported:

    * `Retain`: On any circumstances, retain the backup file when deleting the backup CR.
    * `Delete`: On any circumstances, delete the backup file when deleting the backup CR.

>     * `OnFailure`: If the backup fails, delete the backup file when deleting the backup CR.

Reviewer: s/file/data

    {{< copyable "sql" >}}

    ```sql
    update mysql.tidb set VARIABLE_VALUE = '720h' where VARIABLE_NAME = 'tikv_gc_life_time';
    ```

    After the backup, no matter whether the backup is successful or not, as long as the previous `tikv_gc_lifetime` is smaller than `.spec.tikvGCLifeTime`, TiDB Operator will try to set `tikv_gc_lifetime` to the previous value.

    In extreme cases, if TiDB Operator fails to access the database, TiDB Operator cannot automatically recover the value of `tikv_gc_lifetime` and treats the backup as failed. At this time, you can view `tikv_gc_lifetime` of the current TiDB cluster using the following statement:

    {{< copyable "sql" >}}

    ```sql
    select VARIABLE_NAME, VARIABLE_VALUE from mysql.tidb where VARIABLE_NAME like "tikv_gc_life_time";
    ```

    If the value of `tikv_gc_lifetime` is set too large (usually 10m), you need to set it back to the previous value:

    {{< copyable "sql" >}}

    ```sql
    update mysql.tidb set VARIABLE_VALUE = '10m' where VARIABLE_NAME = 'tikv_gc_life_time';

>     ```

Author: @DanielZhangQD PTAL if these commands are correct.

* `.spec.cleanPolicy`: The clean policy of the backup data when the backup CR is deleted after the backup is completed.

    Three clean policies are supported:

    * `Retain`: On any circumstances, retain the backup data when deleting the backup CR.
    * `Delete`: On any circumstances, delete the backup data when deleting the backup CR.
    * `OnFailure`: If the backup fails, delete the backup data when deleting the backup CR.

    If this field is not configured, or if you configure a value other than the three policies above, the backup data is retained.

>     Note that in v1.1.2 and earlier versions, this field does not exist. The backup data is deleted along with the CR by default. For v1.1.3 or later versions, if you want to keep this behavior, set this field to `Delete`.

Author: `file` -> `data` updated.

* `.spec.metadata.namespace`: The namespace where the `Backup` CR is located.
* `.spec.tikvGCLifeTime`: The temporary `tikv_gc_lifetime` time setting during the backup. Defaults to 72h.

    Before the backup begins, if the `tikv_gc_lifetime` setting in the TiDB cluster is smaller than `spec.tikvGCLifeTime` set by the user, TiDB Operator adjusts the value of `tikv_gc_lifetime` to the value of `spec.tikvGCLifeTime`. This operation makes sure that the backup data is not garbage-collected by TiKV:

    {{< copyable "sql" >}}

    ```sql
    update mysql.tidb set VARIABLE_VALUE = '720h' where VARIABLE_NAME = 'tikv_gc_life_time';

>     ```

Reviewer: No need to add command here as this will be done automatically by TiDB Operator.
